README.md
# pint

## A simple compiler from at subset of Python to IntCode, as defined in [Advent of Code 2019](https://adventofcode.com/2019/day/9). 

## Use

You must have [Dart](https://dart.dev) installed. If you clone this repository, you can compile with:

```
pub run pintc.dart examples/div.pint div.int
```

and then use your own IntCode interpreter to run the generated `div.int` program, or use the one included here, by running:

```
pub run intcode.dart div.int
```

## Caveats

The compiler is written in Dart, but is otherwise entirely (s)elf contained.
It supports functions (with multiple arguments and multiple return values), but
other than that, it is only a small step beyond an assembler. It cannot parse
complicated expressions; most lines of code are one-to-one IntCode instructions
There is currently very little feedback in terms of static analysis, so it is
highly recommended to compile to python (by specifying a `.py` file as output:
`pintc source.pint target.py`) and using tools for python to debug. For most
errors, the compiler will simply reply `Cannot parse:` and then the line number
and line.

## Syntax

Currently, the following constructions are supported:

### Variables

Any sequence of upper and lower case letter are valid variable names.
Currently, variables are global if used outside of a function definition, and
otherwise they are local to the function. Pythons `global` declaration and
implicit globals have not been implemented yet.

### Operations on variables

Addition and multiplication, where the operands can also be (positive/negative) constants

```
a = b + c
a = b * c
```

This includes the shorthand

```
a += b
a *= b
```

Or simply assignment

```
a = b
```

### Input and output

```
a = input()
output(a)
```

Alternatively, to be consistent with generated Python:

```
a = int(input())
print(a)
```

### Blocks

As in Python, code blocks are indentation controlled. `pint` is slightly more
relaxed, but not in a helpful way: Any indented line of code is a child of the
most recent line of code will less indentation. A line without such a parent is
part of the main program.

### Conditionals

Conditional depend on boolean expressions, which are here restricted to two
operands and the operations `<`, `<=`, `>`, `>=`, `==`, and `!=`.

Simple `if`:

```
if a < b:
  max = b
```

Currently, `else` and `elif` is not supported

Simple `while`:

```
while a < b:
  sum += a
  a += 1
```

### Function declaration

```
def max(a, b):
  if a < b:
    return b
  return a
```

The `def` must not be indented.

Recursion is supported, and functions can take any number of arguments and
return values.

Basic `IntCode` does not have a division operator, but with functions, we can write our own:

```
def divide(dividend, divisor):
  if dividend < divisor:
    return 0, dividend
  doubledivisor = divisor * 2
  halfquotient, remainder = divide(dividend, doubledivisor)
  quotient = halfquotient * 2
  if remainder >= divisor:
    divisor *= -1
    remainder += divisor
    quotient += 1
  return quotient, remainder
```

### Ascii support

Constants can in general be replace by a single quoted char, which then takes
the Ascii value of that char (in general, the Utf-16 code of the char). This
means you can write:

```
if digit < '0':
  result *= sign
  return result
digit -= '0'
```

see [examples/ascii.pint](examples/ascii.pint) for more inspiration


CHANGELOG.md
## 1.0.0

Initial version

* Read input into variable
* Write variable or constant to output
* Assignment to variables
* Single addition and multiplication of variables and constants
* Subtraction of constant
* Functions with multiple arguments and multiple return values
* Output to intcode and Python


test/pint_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:pint/pint.dart';
import 'package:pint/intcode_vm.dart';
import 'package:pint/src/util.dart';

List<BigInt> runIntCodeFromFile(String filename, List<BigInt> input) {
  var vm = IntCode(parsePintFile(File(filename)).toIntCode());
  vm.input.addAll(input);
  expect(vm.run(), isFalse);
  return vm.output.toList();
}

void main() {
  test('compile empty', () {
    var intcode = parsePintLines([]).toIntCode();
    expect(intcode, [B(99)]);
  });
  test('add.pint', () {
    var range = B(10);
    for (var a = -range; a <= range; a += BigInt.one) {
      for (var b = -range; b <= range; b += BigInt.one) {
        expect(runIntCodeFromFile('examples/add.pint', [a,b]), [a+b]);
      }
    }
  });
  test('div.pint', () {
    expect(runIntCodeFromFile('examples/div.pint', [B(13),B(5)]), [B(2),B(3)]);
    expect(runIntCodeFromFile('examples/div.pint', [B(-13),B(5)]), [B(-2),B(-3)]);
    expect(runIntCodeFromFile('examples/div.pint', [B(13),B(-5)]), [B(-2),B(3)]);
    expect(runIntCodeFromFile('examples/div.pint', [B(-13),B(-5)]), [B(2),B(-3)]);
    expect(runIntCodeFromFile('examples/div.pint', [B(-23),B(7)]), [B(-3),B(-2)]);
  });
  test('range.pint', () {
    expect(runIntCodeFromFile('examples/range.pint', [B(3), B(9), B(2)]), [B(3), B(5), B(7)]);
    expect(runIntCodeFromFile('examples/range.pint', [B(2), B(9), B(3)]), [B(2), B(5), B(8)]);
  });
  test('swap.pint', () {
    expect(runIntCodeFromFile('examples/swap.pint', [B(1), B(2)]), [B(2), B(1)]);
  });
  test('binom.pint', () {
    expect(runIntCodeFromFile('examples/binom.pint', [B(10), B(5)]), [B(252)]);
  });
  test('hello.pint', () {
    expect(runIntCodeFromFile('examples/hello.pint', []), 'Hello, World!\n'.codeUnits.map((c) => BigInt.from(c)).toList());
  });
}


lib/pint.dart
import 'dart:io';
import 'package:pint/src/pint_ast.dart';

var indentRegExp = RegExp(r'^\s*');

Main parsePintFile(File input) => parsePintLines(input.readAsLinesSync());

Main parsePintLines(List<String> pintSourceLines) {
  var blocks = <Instruction>[];
  var indent = <Instruction, int>{};
  var mainblock = Main();
  indent[mainblock] = -1;
  blocks.add(mainblock);
  var linenum = 0;
  for (var line in pintSourceLines) {
    ++linenum;
    var indentMatch = indentRegExp.firstMatch(line);
    if (indentMatch.end == line.length) {
      continue;
    }
    Instruction ins;
    try {
      ins = Instruction.parse(line);
    } on String catch (e) {
      throw 'Line $linenum: $e\n$line';
    }
    indent[ins] = indentMatch.end;
    while (indent[ins] <= indent[blocks.last]) {
      blocks.removeLast();
    }
    if (!blocks.last.addChild(ins)) {
      throw ('Line $linenum: Unexpected indentation level:\n$line');
    }
    blocks.add(ins);
  }
  return mainblock;
}


lib/src/util.dart
import 'dart:collection';
import 'dart:convert';

T TODO<T>() => throw 'Functionality not implemented.';

BigInt B(int i) => BigInt.from(i);

BigInt B0 = BigInt.zero;
BigInt B1 = BigInt.one;
BigInt BM1 = B(-1);
BigInt B2 = B(2);
BigInt B3 = B(3);
BigInt B4 = B(4);
BigInt B10 = B(10);

extension StreamLines on Stream<List<int>> {
  Stream<String> get lines =>
      transform(Utf8Decoder()).transform(LineSplitter());
}

class DefaultValueMapWrapper<K, V> extends MapBase<K, V> {
  V defaultValue;
  Map<K, V> wrapped;

  DefaultValueMapWrapper(this.wrapped, this.defaultValue);

  @override
  V operator [](Object key) => wrapped[key] ?? defaultValue;

  @override
  void operator []=(Object key, V value) => wrapped[key] = value;

  @override
  Iterable<K> get keys => wrapped.keys;

  @override
  V remove(Object key) => wrapped.remove(key);

  @override
  void clear() => wrapped.clear();
}


lib/intcode_vm.dart
import 'dart:io';
import 'dart:collection';
import 'package:pint/src/util.dart';

class IntCode {
  var input = Queue<BigInt>(), output = Queue<BigInt>();
  Map d;
  var pc = B0, rel = B0;

  IntCode(List<BigInt> prog)
      : d = DefaultValueMapWrapper<BigInt,BigInt>(
            {for (int i = 0; i < prog.length; ++i) BigInt.from(i): prog[i]},
            B0);

  IntCode.fromFile(String filename) {
    var prog = File(filename)
        .readAsLinesSync()[0]
        .split(',')
        .map(BigInt.parse)
        .toList();
    d = DefaultValueMapWrapper(
        {for (int i = 0; i < prog.length; ++i) BigInt.from(i): prog[i]}, B0);
  }

  IntCode.fromStdin() {
    var prog = stdin.readLineSync().split(',').map(BigInt.parse).toList();
    d = DefaultValueMapWrapper(
        {for (int i = 0; i < prog.length; ++i) BigInt.from(i): prog[i]}, B0);
  }

  String disasseble(int op, BigInt a1, BigInt a2, BigInt a3) {
    switch (op) {
      case 1:
        return 'd[$a3] = d[$a1] + d[$a2] = ${d[a1]} + ${d[a2]} = ${d[a1] + d[a2]};';
      case 2:
        return 'd[$a3] = d[$a1] * d[$a2] = ${d[a1]} * ${d[a2]} = ${d[a1] * d[a2]};';
      case 3:
        return 'd[$a1] = input();';
      case 4:
        return 'output(d[$a1]) = output(${d[a1]});';
      case 5:
        return 'if d[$a1] = ${d[a1]} != 0: goto d[$a2] = ${d[a2]};';
      case 6:
        return 'if d[$a1] = ${d[a1]} == 0: goto d[$a2] = ${d[a2]};';
      case 7:
        return 'd[$a3] = d[$a1] < d[$a2]  <=>  ${d[a1]} < ${d[a2]}  <=>  ${d[a1]<d[a2] ? 1 : 0}';
      case 8:
        return 'd[$a3] = d[$a1] == d[$a2]  <=>  ${d[a1]} == ${d[a2]}  <=>  ${d[a1] == d[a2] ? 1 : 0}';
      case 9:
        return 'rel += d[$a1] = ${d[a1]}  =>  rel = ${rel + d[a1]}';
      case 99:
        return 'exit()';
      default:
        return 'UNKNOWN OP: $op';
    }
  }

  bool run([bool verbose = false, bool printmemory = false]) {
    while (true) {
      if (printmemory) {
        stdout.writeln();
        var hi = d.keys.whereType<BigInt>().fold(B0, (a,b)=> a > b ? a : b);
        var base = B0;
        while (base <= hi) {
          stdout.write('  $base:');
          for (var i = B0 ; i < B(5) ; i += B1) {
            stdout.write('\t${d[base+i]}');
          }
          stdout.writeln();
          base += B(5);
        }
        stdout.writeln();
      }
      var op = d[pc].toInt();
      BigInt a1, a2, a3;
      if (op >= 20000) {
        op -= 20000;
        a3 = rel + d[pc + B3];
      } else if (op >= 10000) {
        op -= 10000;
        a3 = pc + B3;
      } else {
        a3 = d[pc + B3];
      }
      if (op >= 2000) {
        op -= 2000;
        a2 = rel + d[pc + B2];
      } else if (op >= 1000) {
        op -= 1000;
        a2 = pc + B2;
      } else {
        a2 = d[pc + B2];
      }
      if (op >= 200) {
        op -= 200;
        a1 = rel + d[pc + B1];
      } else if (op >= 100) {
        op -= 100;
        a1 = pc + B1;
      } else {
        a1 = d[pc + B1];
      }
      if (verbose) {
        stdout.writeln('$pc: ${disasseble(op, a1, a2, a3)}');
      }
      switch (op) {
        case 1:
          d[a3] = d[a1] + d[a2];
          pc += B4;
          break;
        case 2:
          d[a3] = d[a1] * d[a2];
          pc += B4;
          break;
        case 3:
          if (input.isEmpty) return true;
          d[a1] = input.removeFirst();
          pc += B2;
          break;
        case 4:
          output.add(d[a1]);
          pc += B2;
          break;
        case 5:
          pc = d[a1] != B0 ? d[a2] : pc + B3;
          break;
        case 6:
          pc = d[a1] == B0 ? d[a2] : pc + B3;
          break;
        case 7:
          d[a3] = d[a1] < d[a2] ? B1 : B0;
          pc += B4;
          break;
        case 8:
          d[a3] = d[a1] == d[a2] ? B1 : B0;
          pc += B4;
          break;
        case 9:
          rel += d[a1];
          pc += B2;
          break;
        case 99:
          return false;
        default:
          throw 'Unknown op-code: ${op} at $pc';
      }
    }
  }
}


lib/src/pint_ast.dart
import 'util.dart';
import 'dart:math';
import 'dart:convert';

BigInt tryParseBigIntOrChar(String input) {
  var bi = BigInt.tryParse(input);
  if (bi != null) return bi;
  if (input.length < 3 || input[0] != "'" || input[input.length - 1] != "'") return null;
  var js = jsonDecode('"${input.substring(1,input.length-1)}"').toString();
  if (js.length != 1) return null;
  return B(js.codeUnitAt(0));
}

abstract class Instruction {
  Instruction next, parent;
  int codeIndex;
  Instruction();

  Iterable<String> toPython([int indent = 0]);

  factory Instruction.parse(String line) {
    return Op.tryParse(line) ??
        Input.tryParse(line) ??
        Output.tryParse(line) ??
        Conditional.tryParse(line) ??
        Definition.tryParse(line) ??
        Call.tryParse(line) ??
        Return.tryParse(line) ??
        (throw 'Cannot parse:\n$line');
  }

  Iterable<String> get readVariables;
  Iterable<String> get writeVariables;

  bool addChild(Instruction child) => false;

  Iterable<Instruction> get descendants sync* {
    yield this;
  }

  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum);
}

abstract class Block extends Instruction {
  Instruction firstChild, lastChild;

  Iterable<Instruction> get children sync* {
    for (var child = firstChild ; child != null ; child = child.next) {
      yield child;
    }
  }

  @override
  bool addChild(Instruction child) {
    child.parent = this;
    if (firstChild == null) {
      firstChild = lastChild = child;
    } else {
      lastChild.next = child;
      lastChild = child;
    }
    return true;
  }

  @override
  Iterable<Instruction> get descendants sync* {
    yield this;
    for (var child in children) {
      yield* child.descendants;
    }
  }
}

var opRegExp = RegExp(
    r"^(?<indent> *)(?<target>[A-Za-z]+) (?<assign>=|\+=|-=|\*=) (?<left>'.*'|[A-Za-z]+|-?[0-9]+)( (?<op>[-+*]) (?<right>'.*'|[A-Za-z]+|-?[0-9]+))?$");

class Op extends Instruction {
  String target, left, right, op;

  Op(RegExpMatch match) {
    target = match.namedGroup('target');
    left = match.namedGroup('left');
    op = match.namedGroup('op');
    right = match.namedGroup('right');
    var assign = match.namedGroup('assign');
    if (assign.length == 2) {
      right = left;
      op = assign[0];
      left = target;
    }
    if (op == '-') {
      var rval = tryParseBigIntOrChar(right);
      if (rval == null) {
        throw 'Cannot subtract variable:\n${match.input}';
      }
      op = '+';
      right = '${-rval}';
    } else if (op == null) {
      op = '+';
      right = '0';
    }
  }

  factory Op.tryParse(String line) {
    var match = opRegExp.firstMatch(line);
    return match == null ? null : Op(match);
  }

  @override
  Iterable<String> get readVariables => [left, right];

  @override
  Iterable<String> get writeVariables => [target];

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    yield '${''.padLeft(indent)}$target = $left $op $right';
  }

  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
    }
    var code = [op == '+' ? B(1) : B(2), null, null, null];
    var val = tryParseBigIntOrChar(left);
    if (val != null) {
      code[1] = val;
      code[0] += B(100);
    } else {
      var offset = globals[left];
      if (offset != null) {
        code[1] = B(offset);
      } else {
        offset = locals[left];
        if (offset == null) locals[left] = offset = locals.length;
        code[1] = B(offset);
        code[0] += B(200);
      }
    }
    val = tryParseBigIntOrChar(right);
    if (val != null) {
      code[2] = val;
      code[0] += B(1000);
    } else {
      var offset = globals[right];
      if (offset != null) {
        code[2] = B(offset);
      } else {
        offset = locals[right];
        if (offset == null) locals[right] = offset = locals.length;
        code[2] = B(offset);
        code[0] += B(2000);
      }
    }
    var offset = globals[target];
    if (offset != null) {
      code[3] = B(offset);
    } else {
      offset = locals[target] ?? (locals[target] = locals.length);
      code[3] = B(offset);
      code[0] += B(20000);
    }
    return code;
  }
}

var inputRegExp =
    RegExp(r'^(?<indent> *)(?<target>[A-Za-z]+) *= *(input\(\)|int\(input\(\)\))$');

class Input extends Instruction {
  String target;

  Input(RegExpMatch match) {
    target = match.namedGroup('target');
  }

  factory Input.tryParse(String line) {
    var match = inputRegExp.firstMatch(line);
    return match == null ? null : Input(match);
  }

  @override
  Iterable<String> get readVariables => [];

  @override
  Iterable<String> get writeVariables => [target];

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    yield '${''.padLeft(indent)}$target = int(input())';
  }

  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
    }
    var offset = globals[target];
    if (offset != null) {
      return [B(3), B(offset)];
    } else {
      offset = locals[target];
      if (offset == null) {
        locals[target] = offset = locals.length;
      }
      return [B(203), B(offset)];
    }
  }
}

var outputRegExp =
    RegExp(r"^(?<indent> *)(print|output) *\( *(?<arg>'.*'|[A-Za-z]+|-?[0-9]+) *\) *$");

class Output extends Instruction {
  String arg;

  Output(RegExpMatch match) {
    arg = match.namedGroup('arg');
  }

  factory Output.tryParse(String line) {
    var match = outputRegExp.firstMatch(line);
    return match == null ? null : Output(match);
  }

  @override
  Iterable<String> get readVariables => RegExp(r'^[A-Za-z]*$').hasMatch(arg) ? [arg] : [];

  @override
  Iterable<String> get writeVariables => [];

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    yield '${''.padLeft(indent)}print($arg)';
  }

  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
    }
    var val = tryParseBigIntOrChar(arg);
    if (val != null) {
      return [B(104), val];
    } else {
      if (arg[0] == "'") {
        return jsonDecode('"${arg.substring(1,arg.length-1)}"').toString().codeUnits.expand((c) => [B(104), B(c)]).toList();
      } else {
        var offset = globals[arg];
        if (offset != null) {
          return [B(4), B(offset)];
        } else {
          offset = locals[arg];
          if (offset == null) {
            locals[arg] = offset = locals.length;
          }
          return [B(204), B(offset)];
        }
      }
    }
  }
}

var condRegExp = RegExp(
    r"^(?<indent> *)(?<control>if|while) (?<left>'.*'|[A-Za-z]+|-?[0-9]+) (?<op>==|!=|<|>|<=|>=) (?<right>'.*'|[A-Za-z]+|-?[0-9]+):$");

class Conditional extends Block {
  String control, left, right, op;

  Conditional(RegExpMatch match) {
    control = match.namedGroup('control');
    left = match.namedGroup('left');
    op = match.namedGroup('op');
    right = match.namedGroup('right');
  }

  factory Conditional.tryParse(String line) {
    var match = condRegExp.firstMatch(line);
    return match == null ? null : Conditional(match);
  }

  @override
  Iterable<String> get readVariables => [left, right].where((exp) => tryParseBigIntOrChar(exp) == null);

  @override
  Iterable<String> get writeVariables => [];

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    yield ''.padLeft(indent) + 'if $left $op $right:';
    for (var child in children) {
      yield* child.toPython(indent + 2);
    }
  }

  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
    }
    if (op == '>' || op == '<=') {
      op = op == '>' ? '<' : '>=';
      var tmp = left;
      left = right;
      right = tmp;
    }
    var codes = [
      B0,
      tryParseBigIntOrChar(left),
      tryParseBigIntOrChar(right),
      B(globals['_cmp']),
      B(1000),
      B(globals['_cmp']),
      null
    ];
    if (codes[1] != null) {
      codes[0] += B(100);
    } else if (globals[left] != null) {
      codes[1] = B(globals[left]);
    } else {
      codes[0] += B(200);
      codes[1] = B(locals[left] ??= locals.length);
    }
    if (codes[2] != null) {
      codes[0] += B(1000);
    } else if (globals[right] != null) {
      codes[2] = B(globals[right]);
    } else {
      codes[0] += B(2000);
      codes[2] = B(locals[right] ??= locals.length);
    }
    switch (op) {
      case '<':
        codes[0] += B(7);
        codes[4] += B(6);
        break;
      case '>=':
        codes[0] += B(7);
        codes[4] += B(5);
        break;
      case '==':
        codes[0] += B(8);
        codes[4] += B(6);
        break;
      case '!=':
        codes[0] += B(8);
        codes[4] += B(5);
        break;
      default:
        throw 'Conditional $op not implemented';
    }
    for (var child in children) {
      codes.addAll(
          child.encode(locals, globals, functions, opnum + codes.length));
    }
    if (control == 'while') {
      codes.addAll([B(1106), B(0), B(opnum)]);
    }
    codes[6] = B(opnum + codes.length);
    return codes;
  }
}

var funRegExp = RegExp(
    r'^(?<indent>)def (?<name>[A-Za-z]+)\( *(?<args>[A-Za-z]+( *, *[A-Za-z]+)*)? *\):$');

class Definition extends Block {
  String name;
  List<String> args;

  @override
  String toString() => 'def $name(${args.join(', ')}) @ $codeIndex';

  Definition(RegExpMatch match) {
    name = match.namedGroup('name');
    args = match.namedGroup('args')?.split(RegExp(r' *, *'))?.toList() ?? [];
    if (args.length != args.toSet().length) {
      throw 'Parameter names are not unique';
    }
  }

  factory Definition.tryParse(String line) {
    var match = funRegExp.firstMatch(line);
    return match == null ? null : Definition(match);
  }

  @override
  Iterable<String> get readVariables => [];

  @override
  Iterable<String> get writeVariables => [];

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    yield '${''.padLeft(indent)}def $name(${args.join(', ')}):';
    for (var child in children) {
      yield* child.toPython(indent + 2);
    }
  }

  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
      if (globals.keys.toSet().intersection(args.toSet()).isNotEmpty) {
        throw 'Parameters intersect with global variables';
      }
    }
    var returnValueCount = descendants
        .whereType<Return>()
        .map((ins) => ins.expr.length)
        .fold(0, max);
    if (!(lastChild is Return)) {
      // to ensure that execution never overflows function
      addChild(Return.tryParse('return'));
    }
    locals = {
      '_return_to': 0,
      for (int i = 0; i < args.length; ++i) args[i]: i + 1,
      for (int i = 0; i < returnValueCount; ++i)
        '_return_$i': args.length + 1 + i,
    };
    for (var local in descendants.expand((desc) => [...desc.readVariables, ...desc.writeVariables]).toSet()) {
      locals[local] ??= locals.length;
    }
    var codes = <BigInt>[];
    for (var child in children) {
      codes.addAll(
          child.encode(locals, globals, functions, opnum + codes.length));
    }
    return codes;
  }
}

var callRegExp = RegExp(
    r"^(?<indent> *)((?<target>[A-Za-z]+( *, *[A-Za-z]+)*) = )?(?<name>[A-Za-z]+)\((?<args>('.*'|[A-Za-z]+|-?[0-9]+)( *, *('.*'|[A-Za-z]+|-?[0-9]+))*)?\)$");

class Call extends Instruction {
  String name;
  List<String> args, targets;

  Call(RegExpMatch match) {
    targets = match.namedGroup('target')?.split(RegExp(r' *, *'))?.toList() ?? [];
    name = match.namedGroup('name');
    args = match.namedGroup('args')?.split(RegExp(r' *, *'))?.toList() ?? [];
  }

  factory Call.tryParse(String line) {
    var match = callRegExp.firstMatch(line);
    return match == null ? null : Call(match);
  }

  @override
  Iterable<String> get readVariables => args.where((arg) => tryParseBigIntOrChar(arg) == null);

  @override
  Iterable<String> get writeVariables => targets;

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    if (targets.isEmpty) {
      yield '${''.padLeft(indent)}$name(${args.join(', ')})';
    } else {
      yield '${''.padLeft(indent)}${targets.join(', ')} = $name(${args.join(', ')})';
    }
  }

  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
    }
    var codes = <BigInt>[];
    for (var i = 0 ; i < args.length ; ++i) {
      var code = [
        B(21001),
        tryParseBigIntOrChar(args[i]),
        B(0),
        B(locals.length + 1 + i)
      ];
      if (code[1] != null) {
        code[0] += B(100);
      } else {
        if (globals[args[i]] != null) {
          code[1] = B(globals[args[i]]);
        } else {
          code[0] += B(200);
          code[1] = B(locals[args[i]] ??= locals.length);
        }
      }
      codes.addAll(code);
    }
    codes.addAll([B(109), B(locals.length)]);
    codes.addAll([B(21101), B(opnum + codes.length + 7), B(0), B(0)]);
    codes.addAll([B(1106), B(0), B(functions[name]?.codeIndex ?? -1)]);
    codes.addAll([B(109), B(-locals.length)]);
    for (var i = 0 ; i < targets.length ; ++i) {
      var code = [
        B(1201),
        B(locals.length + 1 + i + (functions[name]?.args?.length ?? -1)),
        B(0),
        null
      ];
      if (globals[targets[i]] != null) {
        code[3] = B(globals[targets[i]]);
      } else {
        code[0] += B(20000);
        code[3] = B(locals[targets[i]] ??= locals.length);
      }
      codes.addAll(code);
    }

    return codes;
  }
}

var returnRegExp = RegExp(
    r"^(?<indent> *)return( +(?<expr>('.*'|[A-Za-z]+|-?[0-9]+)( *, *('.*'|[A-Za-z]+|-?[0-9]+))*))? *$");

class Return extends Instruction {
  List<String> expr;

  Return(RegExpMatch match) {
    expr = match.namedGroup('expr')?.split(RegExp(r' *, *')) ?? [];
  }

  factory Return.tryParse(String line) {
    var match = returnRegExp.firstMatch(line);
    return match == null ? null : Return(match);
  }

  @override
  Iterable<String> get readVariables => expr.where((exp) => tryParseBigIntOrChar(exp) == null).toList();

  @override
  Iterable<String> get writeVariables => [];

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    yield '${''.padLeft(indent)}return ${expr.join(', ')}';
  }

  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
    }
    var codes = <BigInt>[];
    for (var i = 0; i < expr.length; ++i) {
      var code = [
        B(21001),
        tryParseBigIntOrChar(expr[i]),
        B(0),
        B(locals['_return_$i'])
      ];
      if (code[1] != null) {
        code[0] += B(100);
      } else {
        if (globals[expr[i]] != null) {
          code[1] = B(globals[expr[i]]);
        } else {
          code[0] += B(200);
          code[1] = B(locals[expr[i]] ??= locals.length);
        }
      }
      codes.addAll(code);
    }
    codes.addAll([B(2106), B(0), B(0)]);
    return codes;
  }
}

class Main extends Block {
  @override
  Iterable<String> get readVariables => [];

  @override
  Iterable<String> get writeVariables => [];

  @override
  Iterable<String> toPython([int indent = 0]) sync* {
    yield '';
    for (var child in children.whereType<Definition>()) {
      yield* child.toPython(indent);
      yield '';
    }
    for (var child in children.where((child) => !(child is Definition))) {
      yield* child.toPython(indent);
    }
    yield '';
  }

  List<BigInt> toIntCode() => encode({}, {}, {}, 0);

  // layout: statements, functions, global variables, stack (local variables)
  @override
  List<BigInt> encode(Map<String, int> locals, Map<String, int> globals,
      Map<String, Definition> functions, int opnum) {
    if (codeIndex == null) {
      codeIndex = opnum;
    } else {
      assert(codeIndex == opnum);
    }
    locals = {'_cmp': 0}; // reserved comparison register
    globals = {'_cmp': 0};
    functions = {for (var func in descendants.whereType<Definition>()) func.name : func};
    var statements = children.where((child) => !(child is Definition)).toList();
    var funcs = children.whereType<Definition>().toList();
    var len = funcs.isNotEmpty ? 2 : 0; // initial stack offset setting
    for (var statement in statements) {
      len += statement.encode(locals, globals, functions, len).length;
    }
    len += 1; // exit code 99
    for (var func in funcs) {
      len += func.encode({}, globals, functions, len).length;
    }
    globals = {for (var e in locals.entries) e.key: e.value + len};
    len += globals.length;
    locals = {};
    var prog = funcs.isNotEmpty ? [B(109), B(len)] : <BigInt>[];
    for (var statement in statements) {
      prog.addAll(statement.encode(locals, globals, functions, prog.length));
    }
    prog.add(B(99));
    for (var function in funcs) {
      prog.addAll(function.encode(locals, globals, functions, prog.length));
    }
    return prog;
  }
}
